libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -0,0 +1,550 @@
+        // calculate intermediate point and scale to margin
+        Vector2f intermediate_pt = (after_pt + before_pt) * 0.5f;
+        float intermediate_len = intermediate_pt.length();
+        intermediate_pt *= (margin_cm / intermediate_len);
+
+        // find final point which is inside the original polygon
+        _polyfence_pts[i] = boundary[i] + intermediate_pt;
+        if (Polygon_outside(_polyfence_pts[i], boundary, num_points)) {
+            _polyfence_pts[i] = boundary[i] - intermediate_pt;
+            if (Polygon_outside(_polyfence_pts[i], boundary, num_points)) {
+                // could not find a point on either side that was within the fence so fail
+                // this can happen if fence lines are closer than margin_m
+                return false;
+            }
+        }
+
+        _polyfence_numpoints++;
+    }
+
+    // last point is same as first point
+    _polyfence_pts[_polyfence_numpoints] = _polyfence_pts[0];
+    _polyfence_numpoints++;
+
+    // record fence update time so we don't process this exact fence again
+    _polyfence_update_ms = fence_ptr->get_boundary_update_ms();
+
+    return true;
+}
+
+// create a visibility graph of the polygon fence
+// returns true on success
+// requires create_polygon_fence_with_margin to have been run
+bool AP_OADijkstra::create_polygon_fence_visgraph()
+{
+    // exit immediately if no polygon fence (with margin)
+    if (_polyfence_numpoints == 0) {
+        return false;
+    }
+
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr == nullptr) {
+        return false;
+    }
+
+    // get polygon boundary
+    uint16_t num_points;
+    const Vector2f* boundary = fence_ptr->get_boundary_points(num_points);
+    if ((boundary == nullptr) || (num_points < 3)) {
+        return false;
+    }
+
+    // clear polygon visibility graph
+    _polyfence_visgraph.clear();
+
+    // calculate distance from each polygon fence point to all other points
+    for (uint8_t i=0; i<_polyfence_numpoints-1; i++) {
+        for (uint8_t j=i+1; j<_polyfence_numpoints-1; j++) {
+            const Vector2f &start1 = _polyfence_pts[i];
+            const Vector2f &end1 = _polyfence_pts[j];
+            Vector2f intersection;
+            // ToDo: calculation below could be sped up by removing unused intersection and
+            //       skipping segments we know are parallel to our fence-with-margin segments
+            if (!Polygon_intersects(boundary, num_points, start1, end1, intersection)) {
+                // line segment does not intersect with original fence so add to visgraph
+                _polyfence_visgraph.add_item({AP_OAVisGraph::OATYPE_FENCE_POINT, i},
+                                              {AP_OAVisGraph::OATYPE_FENCE_POINT, j},
+                                              safe_sqrt(sq(_polyfence_pts[i].x - _polyfence_pts[j].x) + sq(_polyfence_pts[i].y - _polyfence_pts[j].y)));
```

Comment:
Don't think so.. I think "norm(_polyfence_pts[i], _polyfence_pts[j])" would be sqrt(sq(_polyfence_pts[i]) + sq(_polyfence_pts[j]) ... I think it could be the same as norm(_polyfence_pts[i] - _polyfence_pts[j]) which would be the same as (_polyfence_pts[i] - _polyfence_pts[j]).length().  I think perhaps I was looking to avoid the creation of intermediate Vector2f objects but this function is run so rarely it's probably doesn't matter if it's slightly less efficient.